### ISSUE TITLE
Defining general SLSA level for an organization

### ISSUE THREAD

**Issue by royb-legit** [slsa-framework/slsa#286]
Generally speaking, SLSA levels apply to an organization's software artifacts based on its source code repository, build system, provenance characteristics, etc. But what if an organization wants to declare its "holistic" SLSA level, i.e. the level it is **capable** of producing for its artifacts?

Maybe the framework should provide a definition for an organization's general SLSA level. One possible way would be to define it as the minimum level of all their artifacts. But it can be impractical for an organization with a rich history and hundreds of old artifacts which they don't wish to rebuild. Perhaps a more feasible approach would be to define it as the minimum SLSA level of its source and build systems. That way, once an organization had managed to improve its CI/CD environment and its source and build components meet a higher SLSA level, they could now declare they are "SLSA level X capable", meaning any future artifact built by them is going to be at least SLSA X.

**Comment by mlieberman85** [comment]
I do think there's value in asserting that a build system generates artifacts of a certain SLSA level by default. I don't think there's anything that prevents a system today from claiming it generates SLSA artifacts of a certain level by default.

We can't really take a stance on previously built artifacts.

**Comment by royb-legit** [comment]
Thanks @mlieberman85. That's true, but I think there's a point in defining a general SLSA level for an organization. Let's say some vendor puts a SLSA 3 badge on its website's main page. Does it mean that all of its artifacts are SLSA 3? Maybe they are at least SLSA 3? Or maybe on the contrary, it's just saying that it's cabable of producing SLSA 3 artifacts but some of its artifacts are below 3...
I'm just saying maybe this calls for a clear definition.

**Comment by mlieberman85** [comment]
Can you go a bit more into use cases? I think I agree but want to separate out organization vs. build system, etc.

I think for now anyone can make a claim about what their system does and we need to be flexible. If a system claims it by default is SLSA 2 and can build up to SLSA4 with specific types of builds that seems alright. Currently by its nature, SLSA consists of rules that are part of the build systems as well as rules that are part of the governance of the projects that are being built, e.g. two person code review.

**Comment by royb-legit** [comment]
> Can you go a bit more into use cases?
What do you think about the <vendor puts a SLSA 3 badge on its website's main page> use-case? How would you interpret it?

> I think I agree but want to separate out organization vs. build system, etc.
The separation, as I see it, is that an organization might have multiple different build systems, meeting different SLSA levels, ranging from 0 to 4. Such an organization might want to proudly present the L4 badge, implicitly saying "we are capable of producing SLSA 4 artifacts" even when most of their build systems and artifacts are SLSA 0. And that's perfectly fine, as long as the message is clear and there's no confusion as to the meaning.

**Comment by mlieberman85** [comment]
That's the thing, I don't see something as clear as a single badge for SLSA at this point. I think it makes sense to have 2 things, one is bare minimum SLSA and highest achievable SLSA. I think both are valuable. I think about it though in the case of the customer though. How would I trust that badge? Who is certifying that yes this vendor is actually minimum SLSA level 2 and practically can achieve SLSA 4?

**Comment by royb-legit** [comment]
> I think it makes sense to have 2 things, one is bare minimum SLSA and highest achievable SLSA.

Yes, I agree. So maybe for those SLSA-adopting vendors who wish to present their general SLSA capability, it can be nice to add a new type of "SLSA compliant organization" badge - depicting those two levels. WDYT?

> How would I trust that badge? Who is certifying that yes this vendor is actually minimum SLSA level 2 and practically can achieve SLSA 4?

Indeed something that would have to be dealt with, but no different than the current situation, as described in [the levels page](https://slsa.dev/spec/v0.1/levels): `We envision an accreditation process whereby auditors certify that platforms meet the requirements, which consumers can then rely on`. Do you see it otherwise?

**Comment by david-a-wheeler** [comment]
I don't think this is a good idea. If an organization says they are "SLSA level X", can a potential customer be confident that everything the organization maintains is level X? That'd be hard to check, and could easily lead to users getting misled. If they just say "we're capable of doing X", well, I hope so, but users should only care about they actually achieve.

I remember when many organizations said they were "CMM level 3" or "CMMI level 3", but it really only applied to specific parts of their organization, leading to bait-and-switch issues.

Now I *do* think that organizations can do a lot organization-wide to make a specific project easily meet a SLSA level. We should encourage that, but that's not the same thing. At the level of a specific project it's much easier to verify (and fix).

**Comment by royb-legit** [comment]
@david-a-wheeler you're right, regarding the single level being misleading - it was discussed above, and we moved on to the less confusing "min-max" approach. It can benefit vendors that did actually put a lot of effort into raising the SLSA compliance of their entire SDLC. You want to give such vendors the opportunity to say, for instance, "all of our artifacts are between SLSA 2 and SLSA 3" (whether through a badge or any other popular means). Potentially, this can give an idea of how secure the organization is and how resilient it is for supply-chain attacks.
Like I said above, this assumes there's an accreditation body confirming the information, just as it is required when declaring some specific artifact meets a certain SLSA level (currently described as a limitation in the platform).

**Comment by david-a-wheeler** [comment]
@royb-legit - I think the problem though is that "all our artifacts" can be challenging to evaluate, just because the word "all" is a challenge. How can you be sure you have a complete list? How can you really verify that it's really happening?

I'm *all* for reducing work. If we can address those issues well enough so that users can have confidence in the answers, then maybe it'll be fine, but I think we'll need to address them.

**Comment by royb-legit** [comment]
Right, the "all" part isn't practical in most cases, but presenting the fact that your CI/CD env meets a certain SLSA level can be. According to the current description in some pages, the accreditation body which the framework envisions needs to verify that the source and build systems meet the correct level as the vendor claims. Once you have that, you can specify the minimal and maximal levels which are reflected from all relevant systems. It is true that eventually customers will need some specific artifact from that vendor and they will want to verify that artifact meets their desired level (again, via the accreditation body) but the "general" min and max values can bring value.
If we don't want to rely too much on this theoretical accreditation body, there's really no point in this whole idea. (but again, it seems that we already rely on such a body...)

**Comment by MarkLodato** [comment]
I wholeheartedly agree that we should eventually define how to apply SLSA to more than just individual artifacts. In fact, this is what we do within Google&mdash;[BAB](https://cloud.google.com/security/binary-authorization-for-borg) (the SLSA equivalent) applies on a per-service-account basis. We decided to start SLSA with just applying to individual artifacts because that was the easiest thing to define and get agreement on.

However, I'm not sure about the original suggestion to apply to an entire company. As others have said, you will never have *every* artifact meeting the same SLSA level. For example, developers repeatedly build and test changes on their local machine as part of the normal software development process. Clearly they won't meet SLSA, and it shouldn't, yet these binaries would be included in "all". We need much more precision.

Here are some ideas on how SLSA might apply, in roughly increasing difficulty of coming up with a satisfactory definition:

* Artifact: each individual package version has its own level, independent of all others
* Builder: a build service is accredited(?) to produce artifacts at level X, at least under certain conditions. For example, a builder that is fully hermetic, isolated, etc. could meet SLSA 4 if the build enables the hermetic features and the code is fully reviewed.
* Package: a named resource (e.g. PyPI package "pyyaml") has some sort of guarantee that the current and all future versions will meet some SLSA level
* Service account: all of the container images running as this account meet a minimum level. (We'd also need extensions to apply to the service configuration, e.g. pod spec, as well as artifacts loaded at runtime.)
* Machine: all of the software installed on the machine meet a minimum level. (How do you define this? For Kubernetes, is this up to the service account? Does this include the firmware? Many questions remain.)
* Service: the collection of all things in the above categories that can possibly read or modify user data for that service, i.e. the trusted computing base

A lot of open questions and challenges remain. We'll need to provide room for exceptions since it's usually impossible to get to 100% in practice. The list above is most likely incomplete. I'm not even sure we can provide a standard definition for "service" or even "service account".

Still, I'm hopeful that we can make some progress here over time.
